scheduler/src/cook/mesos/task.clj

```diff
@@ -0,0 +1,167 @@
+        sorted-roles (sort-by #(= "*" %) (keys avail))
+        init-state {:remaining-resources avail
+                    :mesos-messages []
+                    :amount-still-needed amount}]
+    (reduce (fn [{:keys [remaining-resources mesos-messages amount-still-needed]
+                  :as state} role-name]
+              (let [amount-avail (or (remaining-resources role-name) 0)
+                    amount-to-take (min amount-still-needed amount-avail)]
+                (if (pos? amount-to-take)
+                  {:remaining-resources
+                   (assoc remaining-resources role-name (- amount-avail amount-to-take))
+                   :mesos-messages
+                   (conj mesos-messages {:name resource-name
+                                         :type :value-scalar
+                                         :role role-name
+                                         :scalar amount-to-take})
+                   :amount-still-needed (- amount-still-needed amount-to-take)}
+                  state)))
+            init-state
+            sorted-roles)))
+
+(defn range-contains?
+  "true iff val is contained by a mesos-style range e.g. {:begin 1 :end 10}"
+  [val mesos-range]
+  (and (<= (:begin mesos-range) val)
+       (>= (:end mesos-range) val)))
+
+(defn role-containing-port
+  "in: {\"*\" [{:begin 201 :end 202}] \"cook\" [{:begin 203 :end 204}]}, 201
+  ;; out \"*\""
+  [available-ports port]
+  (->> available-ports
+       seq
+       (filter #(some (partial range-contains? port) (val %)))
+       ffirst))
+
+(defn take-ports
+  "Given a set of available resources (in the format returned by resources-by-role),
+  returns a vector of mesos messages that will reserve the specified ports.
+  Note: unlike with scalar-resources, available-resources don't need to be
+  changed by this function, because Fenzo has already provided us with
+  specific ports to use for every task."
+  [available-resources ports-needed]
+  (mapv (fn [port]
+          {:name "ports"
+           :type :value-ranges
+           :role (role-containing-port available-resources port)
+           :ranges [{:begin port :end port}]})
+        ports-needed))
+
+(defn add-ports-to-task-info
+  "Given a set of tasks and offers that were matched together by Fenzo,
+   assigns the specific ports requested by role to each task.
+   Returns the input tasks, decorated with :ports-resource-messages"
+  [available-resources tasks]
+  (let [available-ports (available-resources "ports")]
+    (map (fn [task]
+           (let [ports (:ports-assigned task)
+                 port-env-vars (into {} (map-indexed (fn [i p] [(str "PORT" i) (str p)])
+                                                     ports))]
+             (-> task
+                 (assoc :ports-resource-messages (take-ports available-ports ports))
+                 (update-in [:command :environment] merge port-env-vars))))
+         tasks)))
+
+(defn add-scalar-resources-to-task-info
```

Comment:
I think it is worthwhile to write this function without hardcoding mem and cpus as I expect gpus and disk (among others) will be added soon.